### PR TITLE
make it possible to cancel window functions on ctrl+c

### PR DESCRIPTION
**Pull request by akuzm**
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: tests/queries/0_stateless/01572_kill_window_function.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+query_id="01572_kill_window_function"
+$CLICKHOUSE_CLIENT --query_id="$query_id" --query "SELECT count(1048575) OVER (PARTITION BY intDiv(NULL, number) ORDER BY number DESC NULLS FIRST ROWS BETWEEN CURRENT ROW AND 1048575 FOLLOWING) FROM numbers(255, 1048575)" >/dev/null 2>&1 &
+client_pid=$!
+wait_for_query_to_start "$query_id"
+$CLICKHOUSE_CLIENT --query "kill query where query_id = '$query_id' and current_database = currentDatabase() format Null"
+wait $client_pid || echo $?
+
+# We have tested for Ctrl+C.
+# The following client flags don't cancel, but should: --max_execution_time,
```
Comment (Algunenano) [inline tests/queries/0_stateless/01572_kill_window_function.sh:22]: The telegram link isn't public, but likely related to this: https://github.com/ClickHouse/ClickHouse/issues/26554 / https://github.com/ClickHouse/ClickHouse/pull/26675#issuecomment-887297402

Comment (akuzm) [reply tests/queries/0_stateless/01572_kill_window_function.sh:22]: > The telegram link isn't public, but likely related to this: #26554 / [#26675 (comment)](https://github.com/ClickHouse/ClickHouse/pull/26675#issuecomment-887297402)

Yep, was on private group. Basically we were talking that we should estimate the query execution limits asynchronously around [here](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Executors/PipelineExecutor.cpp#L803), and then the threads could very cheaply check for `is_cancelled` flag.

Comment (Algunenano) [reply tests/queries/0_stateless/01572_kill_window_function.sh:22]: Thanks for sharing!

File: src/Processors/Transforms/WindowTransform.cpp [hunk 2]
```diff
@@ -1166,6 +1166,23 @@ void WindowTransform::appendChunk(Chunk & chunk)
             // Write out the aggregation results.
             writeOutCurrentRow();
 
+            if (isCancelled())
```
Comment (kitaisreal) [inline src/Processors/Transforms/WindowTransform.cpp:1169]: It definitely will be devirtualized.

**Review by kitaisreal — APPROVED** [review]

**Comment by akuzm** [comment]
Our CI is completely dead but I don't think it is related to this PR.
Also one window function perf test has regressed 5% for Mysterious Reasons, probably some boundary effect in code generation.

File: src/Processors/Transforms/WindowTransform.cpp (hunk 2 above)
Comment (akuzm) [reply src/Processors/Transforms/WindowTransform.cpp:1169]: Btw note how harder it is to reason about this one-line virtual function compared to just a data field. I had to remember that it is virtual and add `final`. "Object-oriented" features and performance are often at odds...

**Pull request merged**